newIDE/app/public/external/piskel/piskel-main.js

```diff
@@ -122,48 +123,64 @@ const loadImagesIntoPiskel = externalEditorInput => {
-        maxHeight = Math.max(image.height, maxHeight);
-        callback();
-      };
-
-      // onload will fire after `src` is set
-      try {
-        image.src = resource.dataUrl;
-      } catch (error) {
-        // Unable to load the image, ignore it.
-        console.error('Unable to load ', resource, error);
-        callback();
-      }
-    },
-    err => {
-      // Finally load the image objects into piskel
-      const piskelFile = pskl.service.ImportService.prototype.createPiskelFromImages_(
-        imageData,
-        externalEditorInput.name,
-        maxWidth,
-        maxHeight,
-        false
-      );
-      piskelController.setPiskel(piskelFile, {});
-
-      // Add original path variable to imported frame objects, so we can overwrite them later when saving changes
-      const layer = piskelController.getLayerAt(0);
-      for (let i = 0; i < piskelController.getFrameCount(); i++) {
-        layer.getFrameAt(i).originalLocalFilePath =
-          externalEditorInput.resources[i].localFilePath;
-        layer.getFrameAt(i).originalResourceName =
-          externalEditorInput.resources[i].name;
-        layer.getFrameAt(i).originalIndex = i;
-      }
-    }
+
+  // Load all the images into a DOM `Image`.
+  await Promise.all(
+    externalEditorInput.resources.map(resource => {
+      return new Promise(resolve => {
+        let hasAlreadyLoadedOrErrored = false;
+
+        const image = new Image();
+        image.onload = () => {
+          if (hasAlreadyLoadedOrErrored) return;
+          hasAlreadyLoadedOrErrored = true;
+
+          imageData.push(image);
+          maxWidth = Math.max(image.width, maxWidth);
+          maxHeight = Math.max(image.height, maxHeight);
+          resolve();
+        };
+        image.onerror = event => {
+          if (hasAlreadyLoadedOrErrored) return;
+          hasAlreadyLoadedOrErrored = true;
+          console.error('Unable to load ', resource, event);
+          resolve();
+        };
+
+        try {
+          // `onload` (or `onerror`) will fire after `src` is set.
+          image.src = resource.dataUrl;
+        } catch (error) {
+          if (hasAlreadyLoadedOrErrored) return;
+          hasAlreadyLoadedOrErrored = true;
+
+          // Unable to load the image, ignore it.
+          console.error('Unable to load ', resource, error);
+          resolve();
+        }
+      });
+    })
   );
+
+  // Finally load the `Image` objects into Piskel.
+  const piskelFile = pskl.service.ImportService.prototype.createPiskelFromImages_(
+    imageData,
+    externalEditorInput.name,
+    maxWidth,
+    maxHeight,
+    false
+  );
+  piskelController.setPiskel(piskelFile, {});
+
+  // Add original path variable to imported frame objects, so we can overwrite them later when saving changes.
+  const layer = piskelController.getLayerAt(0);
+  for (let i = 0; i < piskelController.getFrameCount(); i++) {
+    layer.getFrameAt(i).originalLocalFilePath =
```

Comment:
you can probably put the frame in a variable to avoid calling it 3 times

Reply:
Done